job_definitions/rotate_api_tokens.yml

```diff
@@ -0,0 +1,92 @@
+{% set stages = [{'name': 'Add new tokens', 'command': 'add-new'}, {'name': 'Remove old tokens', 'command': 'remove-old'}] %}
+{% set app_groups = [['search-api'], ['api'], ['admin-frontend', 'briefs-frontend', 'brief-responses-frontend', 'buyer-frontend', 'supplier-frontend', 'user-frontend']] %}
+---
+- job:
+    name: rotate-api-tokens
+    display-name: "Rotate API tokens"
+    project-type: pipeline
+    description: Rotate Data and Search API tokens for a given environment.
+    disabled: false
+    concurrent: false
+    parameters:
+      - choice:
+          name: STAGE
+          choices:
+            - preview
+            - staging
+            - production
+    pipeline:
+      script: |
+        def notify_slack(icon, status) {
+          build job: "notify-slack",
+          parameters: [
+            string(name: 'USERNAME', value: "update_api_tokens"),
+            string(name: 'ICON', value: icon),
+            string(name: 'JOB', value: "Rotate API tokens on ${STAGE}"),
+            string(name: 'CHANNEL', value: "#dm-release"),
+            text(name: 'STAGE', value: "${STAGE}"),
+            text(name: 'STATUS', value: status),
+            text(name: 'URL', value: "<${BUILD_URL}consoleFull|${BUILD_DISPLAY_NAME}>")
+          ]
+        }
+
+{% for stage in stages %}
+        stage("{{ stage.name }}") {
+{% if loop.index == 1 %}
+          currentBuild.displayName = "#${BUILD_NUMBER} - ${STAGE}"
+{% endif %}
+          node {
+            try {
+              sh('docker run -e GITHUB_ACCESS_TOKEN digitalmarketplace/scripts scripts/rotate-api-tokens.sh "{{ stage.command }}" "${STAGE}"')
+
+              notify_slack(':spinner:', '{{ stage.name }} - SUCCESS')
+
+            } catch(e) {
+              notify_slack(':kaboom:', '{{ stage.name }} - FAILED')
+              echo "Error caught"
+              currentBuild.result = 'FAILURE'
+              return
+            }
+          }
+        }
+
+        stage('Deploy') {
+          milestone()
+          input(message: "Do not continue until the Pull Request updating API tokens has been merged to master.")
+          milestone()
+
+{% for app_group in app_groups %}
+          parallel(
+{% for app in app_group %}
+            "{{ app }}": {
+              stage("{{ app }}") {
+                build job: "release-app-paas", parameters: [
+                  string(name: "STAGE", value: "${STAGE}"),
+                  string(name: "APPLICATION_NAME", value: "{{ app }}")
+                ]
+              }
+            },
+{% endfor %}
+          )
+{% endfor %}
+        }
+
+{% if loop.index == 1 %}
```

Comment:
On a small note, could we indent our loops so we can see the logic flow based on indentation? (I know we don't always do this, myself included but I think we should change that)

Reply:
We generally don't indent them for a reason - jinja preserves the whitespace and it leads to the output being indented all over the place. For things where whitespace matters (e.g. yaml), this makes things break.

Reply:
Would some of the jinja control characters fix that? https://stackoverflow.com/questions/35775207/remove-unnecessary-whitespace-from-jinja-rendered-template

Reply:
I tried using these before with little success. I can try again if you think it's an isssue.

Reply:
Don't spend any more than 2 minutes, it's not a game changer